### PR TITLE
[int-set] Fix several bugs in IntSet found by the fuzzer:

### PR DESCRIPTION
**Pull request by garretrieger**
- Fix IntSet::len() for inverted sets with domains smaller than u32.
- Fix overflow bug in BitPage iter: next() was casting an i32 (backwards_index) to u32 prior to comparison. If backwards_index is -1 this leads to unexpected behaviour.
- Fix iter_ranges() on discontinous domains: consider two values adjacent if they are adjancent in the ordered_values() iterator instead of requiring them to be adjacent in the u32 domain.
- Fix iter_ranges() on inclusive sets w/ discontinous domains. Adjacent ranges need to be merged together.

### CODE REVIEW

File: int-set/src/lib.rs [hunk 1]
```diff
@@ -576,14 +589,20 @@ where
...
     AllValuesIter: Iterator<Item = u32>,
+    T: Domain<T>,
 {
     Inclusive {
         ranges: InclusiveRangeIter,
     },
+    InclusiveDiscontinous {
+        ranges: InclusiveRangeIter,
+        current_range: Option<RangeInclusive<u32>>,
+        phantom: PhantomData<T>,
```
Comment (cmyr) [inline int-set/src/lib.rs:604]: why is this necessary?

**Review by cmyr — APPROVED** [review]
A couple notes online but thanks for splitting this out?

File: int-set/src/lib.rs (hunk 1 above)
Comment (garretrieger) [reply int-set/src/lib.rs:594]: The InclusiveDiscontinous arm needs access to the T type inside of the Iterator::next() implementation. The only way I could find to get it the T type was to make it part of RangeIter, from there the phantom data is needed to avoid an unused type parameter error.
